### DISCUSSION TITLE
How to check if Hermes bytecode bundle contains debug info?

### DISCUSSION THREAD

**Discussion by krystofwoldrich** [facebook/hermes#1129] [Q&A] [answered]
When `hermesc` emits a source map, debug information is stripped from the bytecode. 

Is it possible to check in runtime in React Native or Hermes API if the loaded bundle/bundles have debug info present or not?

**Comment by tmikov** [comment]
Hmm, we haven't planned for this, but I think you could do it by calling `HermesInternal.getFunctionLocation()`.

t1.js:
```JavaScript
function foo() {}
print(JSON.stringify(HermesInternal.getFunctionLocation(foo)));
```

Then I tried these commands:
```
$ hermes -emit-binary -out t1.hbc t1.js

$ hermes t1.hbc
{"isNative":false,"segmentID":0,"virtualOffset":72,"fileName":"t1.hbc"}

$ hermes -emit-binary -g -out t1.hbc t1.js

$ hermes t1.hbc
{"isNative":false,"lineNumber":1,"columnNumber":1,"fileName":"t1.js"}
```

See how in one case you get a virtualOffset and in the other line and column?

**Comment by krystofwoldrich** [comment]
Thanks for the reply, this looks very promising. But sadly it seems to help differ only between `g3` and others.

I'm looking for a way to differ between `hermes -emit-binary -out t1.hbc test.js` and `hermes -emit-binary -output-source-map -out t1.hbc test.js`, but in these cases, the function location has the `virtualOffset`.

Based on the bytecode content I assume `--output-source-map` equals `g0` and the first command is `g1` or `g2`, I don't know what is the default.

To be honest, I'm surprised that the error stack trace and `getFunctionLocation` have different outputs. The `t1.hbc` shows the js file location, but `getFunctionLocation` shows the bytecode location.

### test.js
```
print(JSON.stringify((new Error("Test")).stack));
```

### Output
```
$ hermes -emit-binary -out t1.hbc test.js
$ hermes t1.hbc
"Error: Test\n    at global (test.js:1:32)"

$ hermes -emit-binary -output-source-map -out t2.hbc test.js
$ hermes t2.hbc
"Error: Test\n    at global (address at t2.hbc:1:41)"
```

In practice RN app won't have the application code on line 1, so just based on the line number, it's possible to guess.

**Comment by tmikov** [comment, marked as answer]
This all actually has logic behind it.

`-g0` should never be used. It is there for completeness and testing. It emits no location information of any kind. 

`-g1`, which is the default, emits only enough information to generate stack traces. This information is restricted only to locations where an exception can be thrown (this is done to minimize the size of the bundle). Naturally, with `-g1` there is no information about locations of function declarations. `-g1` produces the default behavior one expects from a JS engine - stack traces with locations.

`-g2` emits information for all locations. If you use `-g2`,  `getFunctionLocation()` will return line and column. Of course, the bundle will be larger.

`-g3` is like `-g2` but additionally it emits information making it possible to actually debug, specifically variable names, scopes, etc. Obviously it should not be used in production.

`-output-source-map` should be used when one wants to go to extreme lengths to decrease the size of the bundle, by relying on additional external infrastructure to decode locations. It emits the same information as `-g2`, but instead of placing it in the bundle. it puts it in a source map, which maps from bytecode offset to source offset. Stack traces and `getFunctionLocation()` will return byte offsets. These offsets can be converted to locations by using an external tool and the generated source map, stored somewhere on a server.

As you noted, if you want to differentiate between `-g1` and `-output-source-map` from within the bundle, you can construct an instance of `Error()` and examine its `.stack` property. I should have thought of it earlier, sorry!

Unfortunately I have also become aware that at least some Hermes users, perhaps unintentionally, are using CodePush to deploy source code OTA, which makes the situation really messy. Source code will obviously make the app really slow, but might also screw the result of the above trick.

**Reply by krystofwoldrich** [reply]
Thanks for the details. We [used the stack](https://github.com/getsentry/sentry-react-native/pull/3290/files#diff-4429cfe22307a32be37115ef628c944e05d307b26bcf90211b88a297a6f269a8R99-R109) to determinate `-g1` and `-output-source-map`.

**Reply by tmikov** [reply]
If you use a combination of this trick and `getFunctionLocation()`, you can distinguish between `-g1` and `-g2` 🤣

**Reply by krystofwoldrich** [reply]
Luckily we only need to diff `-g0` and `-g1` to know what source map will work.
